clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -17632,8 +17632,20 @@ Value *CodeGenFunction::EmitAMDGPUBuiltinExpr(unsigned BuiltinID,
   case AMDGPU::BI__builtin_amdgcn_mov_dpp:
   case AMDGPU::BI__builtin_amdgcn_update_dpp: {
     llvm::SmallVector<llvm::Value *, 6> Args;
-    for (unsigned I = 0; I != E->getNumArgs(); ++I)
-      Args.push_back(EmitScalarExpr(E->getArg(I)));
+    for (unsigned I = 0; I != E->getNumArgs(); ++I) {
+      llvm::Value *Arg = EmitScalarExpr(E->getArg(I));
+      // Except first two input operands, all other are imm operands for dpp
+      // intrinsic.
+      if (llvm::is_contained(std::initializer_list<unsigned>{2, 3, 4, 5}, I)) {
```

Comment:
You don't need to hard code this, referring to the generic builtin handling here:
https://github.com/llvm/llvm-project/blob/5f5f82af966e6edcc72df02b36fb54401ab76266/clang/lib/CodeGen/CGBuiltin.cpp#L5716

You can just lookup which arguments require the immediate. We don't actually have unique needs here, the only reason we probably custom emit this is to emit the type to mangle the builtin. Is there a way to let the generic builtin code deal with the argument list?

Reply:
Also if we have this issue here, we have it for every other builtin requiring constant arguments. It would be best to fix all of them

Reply:
How about doing it in a separate patch for all the builtins and keeping this only for dpp?

Reply:
Yes, the others can be fixed separately.